File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        i, j = 0, len(height) - 1
        l_max, r_max = 0, 0
        res = 0

        while i < j:
            if height[i] <= height[j]:
                if l_max < height[i]:
                    l_max = height[i]
                else:
                    res = res + l_max - height[i]
                i += 1
            else:
                if r_max < height[j]:
                    r_max = height[j]
                else:
                    res = res + r_max - height[j]
                j -= 1
        return res
        

File: 0042-trapping-rain-water/0042-trapping-rain-water.js
/**
 * @param {number[]} height
 * @return {number}
 */
var trap = function(height) {
    let i = 0;
    let j = height.length - 1;
    let leftMax = 0;
    let rightMax = 0;
    let res = 0;

    while (i < j) {
        if (height[i] <= height[j]) {
            if (leftMax < height[i]) {
                leftMax = height[i];
            } else {
                res += (leftMax - height[i]);
            }
            i++;
        } else {
            if (rightMax < height[j]) {
                rightMax = height[j];
            } else {
                res += (rightMax - height[j]);
            }
            j--;
        }
    }
    return res;
};